contracts/starknet/jug.cairo

```diff
@@ -0,0 +1,433 @@
+    ilk: felt, what: felt, data: Uint256
+) {
+    auth();
+
+    let (timestamp) = get_block_timestamp();
+
+    let (i) = _ilks.read(ilk);
+
+    // require(block.timestamp == ilks[ilk].rho, "Jug/rho-not-updated");
+    with_attr error_message("End/file-unrecognized-param") {
+        assert timestamp = i.rho;
+    }
+
+    // else revert("Jug/file-unrecognized-param");
+    with_attr error_message("End/file-unrecognized-param") {
+        assert what = 'duty';
+    }
+
+    // if (what == "duty") ilks[ilk].duty = data;
+    _ilks.write(ilk, Ilk(duty=data, rho=i.rho));
+
+    // emit File(ilk, what, data);
+    File_duty.emit(ilk, what, data);
+    return ();
+}
+
+// function file(bytes32 what, uint256 data) external auth {
+@external
+func file_base{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    what: felt, data: Uint256
+) {
+    auth();
+
+    let (timestamp) = get_block_timestamp();
+
+    // else revert("Jug/file-unrecognized-param");
+    with_attr error_message("End/file-unrecognized-param") {
+        assert what = 'base';
+    }
+
+    // if (what == "base") base = data;
+    _base.write(data);
+
+    // emit File(what, data);
+    File_base.emit(what, data);
+    return ();
+}
+
+// function file(bytes32 what, address data) external auth {
+@external
+func file{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(what: felt, data: felt) {
+    auth();
+
+    // else revert("Jug/file-unrecognized-param");
+    with_attr error_message("End/file-unrecognized-param") {
+        assert what = 'vow';
+    }
+
+    // if (what == "vow") vow = data;
+    _vow.write(data);
+
+    // emit File(what, data);
+    File.emit(what, data);
+    return ();
+}
+
+// // --- Stability Fee Collection ---
+//     function drip(bytes32 ilk) external returns (uint256 rate) {
+@external
+func drip{
+    syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr, bitwise_ptr: BitwiseBuiltin*
+}(ilk: felt) -> (rate: Uint256) {
+    alloc_locals;
+
+    let (vat) = _vat.read();
+    let (vow) = _vow.read();
+    let (i) = _ilks.read(ilk);
+    let (base) = _base.read();
+    let (timestamp) = get_block_timestamp();
+
+    // (, uint256 prev) = vat.ilks(ilk);
+    let (_, prev: Uint256) = VatLike.ilks(vat, ilk);
+    // rate = _rpow(base + ilks[ilk].duty, block.timestamp - ilks[ilk].rho, RAY) * prev / RAY;
+    let (x) = add(base, i.duty);
+    let n = timestamp - i.rho;
+    let (un) = _felt_to_uint(n);
+    let (rate) = _rpow(x, un, Uint256(RAY, 0));
+    let (rate) = mul(rate, prev);
+    let (rate) = div(rate, Uint256(RAY, 0));
+
+    // vat.fold(ilk, vow, _int256(rate) - _int256(prev));
+    let (value) = _sub(rate, prev);
+    VatLike.fold(vat, ilk, vow, value);
+
+    // ilks[ilk].rho = block.timestamp;
+    _ilks.write(ilk, Ilk(duty=i.duty, rho=timestamp));
+
+    // emit Drip(ilk);
+    Drip.emit(ilk);
+
+    return (rate=prev);
+}
+
+namespace internal {
```

Comment:
Not sure why do you want to have namespace here? And why not move rpow here?

Reply:
yup it could be moved, namespace would just be for clarity and not expose it elsewhere.